File: src/optimizer/SnNonlcon.m
% This function is to evaluate all nonlinear constraints and their gradients.
% The output c is a column vector, where every entry c(t) indicates a constraint c(t) <= 0.
% ceq is similar, where every entry ceq(t) indicates ceq(t) == 0.
% dc and dceq are gradient matrices, where each row dc(t, :) indicates the partial derivatives
% of c(t) with respect to every optimizable variable (there are param_manager.num_input many).
% dceq is similar.
% The input x is a column vector.
% The end-hook's behavior depends on expinfo.is_refine_mode: in the refining mode,
% precision requirement is higher.
% There are few constraints that depends on expinfo.obj_mode, which are handled in Workspace,
% e.g., in the 'alpha' mode we require omega == 2, etc.

function [c, ceq, dc_, dceq_] = SnNonlcon(x)
  global param_manager workspace;
  param_manager.SetValue(x');
  [gc, gceq] = workspace.Evaluate();  % gc, gceq are cell arrays of GVar
  [c, dc] = param_manager.PackResults(gc);
  [ceq, dceq] = param_manager.PackResults(gceq);
  c = full(c);
  ceq = full(ceq);
  if nargout > 2
    dc_ = dc;
    dceq_ = dceq;
  end
  SnNonlconEndHook(x, c, ceq);
end

File: fast_mm_python/src/optimizer/SnNonlcon.py
"""
Nonlinear constraint evaluation for the optimization problem.
"""

import numpy as np


def SnNonlcon(x, param_manager=None, workspace=None):
    """
    Evaluate nonlinear constraints and their gradients.
    
    Args:
        x: Parameter vector
        param_manager: ParamManager instance (will use global if None)
        workspace: Workspace instance (will use global if None)
    
    Returns:
        c: Inequality constraint values (c <= 0)
        ceq: Equality constraint values (ceq == 0)
        dc: Inequality constraint gradients
        dceq: Equality constraint gradients
    """
    if param_manager is None:
        from ..autograd import get_param_manager
        param_manager = get_param_manager()
    
    if workspace is None:
        from ..evaluation import get_workspace
        workspace = get_workspace()
    
    # Set current parameter values
    param_manager.SetValue(x)
    
    # Evaluate constraints
    gc, gceq = workspace.Evaluate()
    c, dc = param_manager.PackResults(gc)
    ceq, dceq = param_manager.PackResults(gceq)
    
    # Convert to dense arrays if needed
    if hasattr(dc, 'toarray'):
        dc = dc.toarray()
    if hasattr(dceq, 'toarray'):
        dceq = dceq.toarray()
    
    return c, ceq, dc, dceq

